### ISSUE TITLE
:flat is very slow for large file lists

### ISSUE THREAD

**Issue by sammoth** [hut/ranger#379]
Hi,
The :flat feature seems to have some growth behaviour that's making it unusable for large folders. In a folder with ~20 subfolders and ~30k files, running :flat -1 brings ranger to a complete standstill. The progress spinner soon stops visibly animating.

**Comment by amerlyq** [comment]
Are you tried ranger on separate flat directory with 30k files?
For me it seems issue not in `:flat` algorithm but in speed of ranger files fetching itself.
Especially on HDD.
Seems like there need some lazy loading mechanics, wich loads ext info/attributes only for displayed on screen files, to improve this somehow.

**Comment by sammoth** [comment]
I tried ranger on a directory structure like folder1/folder2, where folder 1 contains no other files and folder2 contains about 17k images.

Browsing normally into folder2, ranger takes ~1 minute to load the folder, with the spinner spinning at full speed. Once the folder is loaded, scrolling up and down the files is fully responsive.

Running flat -1 inside folder1, ranger takes more like 5 minutes + to load the folder, with the spinner stuttering, and once the folder is loaded scrolling up and down the folder is slow, taking about 5 seconds to move up or down for each entry in the file list.

**Comment by andya9** [comment]
I can confirm this too. I wonder if using sets instead of lists (or is it already like that?) would help performance; we can use sets because file names are guaranteed to be unique

**Comment by mik13ST** [comment]
Yes, there is some kind unnecessary UI rendering going on because switching to TaskView (W key) makes the loading of the flattened directory (flattening) kind of infinitely faster.

I did some profiling and I can confirm the UI is being continually refreshed and that most time is wasted there. I think the `directory.load_bit_by_bit()` is updating something which causes the constant redrawing of the UI. I might be wrong though.

**Comment by toonn** [comment]
Interesting. Maybe we can come up with a way to delay refreshing the UI when triggering flat?

**Comment by stevium** [comment]
This would be such a useful feature if it only had some lazy loading. 

I tried to chain it with `filter` and `filter_stack` to exclude for example `node_modules` or `dist` folders, but it does not get any better.

I might be missing something, but it seems like the only option we are left with when searching through bigger directories is using `fzf_select`. But it can only select and open one file at a time, without actually having the possibility to preview/manipulate the resulting list.

**Comment by toonn** [comment]
There's also the fd integration on the wiki. That allows you to jump through results in multiple directories with `n/N`, which'd allow you to preview them.

**Comment by stevium** [comment]
Thank you! Didn't know of this amazing command! 
It's exactly what I was trying to force `flat` and `filter` to do, but unique in its own way that it preserves the directory structure while scanning through search results. 
Of course, `flat` is still a preferable approach in case you want to manipulate results in one list, having all the files and paths at your disposal.
